servicetalk-grpc-internal/src/main/java/io/servicetalk/grpc/internal/GrpcStatusUtils.java

```diff
@@ -0,0 +1,162 @@
+        final CharSequence message = headers.get(GRPC_STATUS_MESSAGE);
+        if (message == null) {
+            return null;
+        }
+
+        final byte[] messageBytes = message.toString().getBytes(StandardCharsets.UTF_8);
+        for (int i = 0; i < messageBytes.length; i++) {
+            byte b = messageBytes[i];
+            if (b < ' ' || b >= '~' || (b == '%' && i + 2 < messageBytes.length)) {
+                return decodeMessage(messageBytes);
+            }
+        }
+        return message;
+    }
+
+    /**
+     * Decodes the {@link CharSequence} removing the percent-encoding where needed.
+     *
+     * @param messageBytes the message to decode.
+     * @return the deocded message.
+     */
+    private static CharSequence decodeMessage(final byte[] messageBytes) {
+        ByteBuffer buf = ByteBuffer.allocate(messageBytes.length);
+        for (int i = 0; i < messageBytes.length;) {
+            if (messageBytes[i] == '%' && i + 2 < messageBytes.length) {
+                try {
+                    buf.put((byte) Integer.parseInt(new String(messageBytes, i + 1, 2, StandardCharsets.US_ASCII), 16));
+                    i += 3;
+                    continue;
+                } catch (NumberFormatException e) {
+                    // ignore, fall through, just push the bytes.
+                }
+            }
+            buf.put(messageBytes[i]);
+            i += 1;
+        }
+        return new String(buf.array(), 0, buf.position(), StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Describes the character ranges which need escaping (essentially non-printable characters).
+     *
+     * @param b the character to check.
+     * @return true if it needs escaping, false otherwise.
+     */
+    private static boolean isEscapingChar(byte b) {
+        return b < ' ' || b >= '~' || b == '%';
+    }
+
+    /**
+     * Performs encoding of the message by using a percent-encoding scheme.
+     *
+     * @param msgBytes the encoded message.
+     * @param ri the reader index previously iterated to.
+     * @return the encoded message.
+     */
+    private static CharSequence encodeMessage(byte[] msgBytes, int ri) {
+        byte[] escapedBytes = new byte[ri + (msgBytes.length - ri) * 3];
+        // copy over the good bytes
+        if (ri != 0) {
+            System.arraycopy(msgBytes, 0, escapedBytes, 0, ri);
+        }
+        int wi = ri;
+        for (; ri < msgBytes.length; ri++) {
+            byte b = msgBytes[ri];
+            // Manually implement URL encoding, per the gRPC spec.
+            if (isEscapingChar(b)) {
+                escapedBytes[wi] = '%';
+                escapedBytes[wi + 1] = HEX[(b >> 4) & 0xF];
+                escapedBytes[wi + 2] = HEX[b & 0xF];
+                wi += 3;
+                continue;
+            }
+            escapedBytes[wi++] = b;
+        }
+        return new String(escapedBytes, 0, wi, StandardCharsets.UTF_8);
```

Comment:
If you make a string with the UTF_8 charset we'll get corruption if any of the bytes where > 127. I believe that what we're trying to do here is smuggle a `byte[]` as a `CharSequence`. We typically use the ISO_8859_1 charset for that because it is a single byte charset that will roundtrip any single byte nicely. You can see the netty 'encoding' of the string [here](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/HpackEncoder.java#L310) where we convert it back to bytes.
```suggestion
        return new String(escapedBytes, 0, wi, StandardCharsets.ISO_8859_1);
```

Reply:
After we add more messages for `ProtocolCompatibilityTest` we can see what is the difference between grpc-java and these recommendations. Then we can decide what is better: to be compatible or handle more cases.